OpenRA.Mods.Common/Traits/World/ClusterManager.cs

```diff
@@ -0,0 +1,314 @@
+
+		protected readonly ClusterContents emptyCluster;
+		protected readonly List<ClusterContents> clusters = new List<ClusterContents>();
+		protected readonly World world;
+		protected readonly Map map;
+
+		public ClusterManager(World world)
+		{
+			this.world = world;
+			map = world.Map;
+			clusterCellLayer = new CellLayer<ClusterContents>(world.Map);
+			emptyCluster = new ClusterContents(true);
+			clusters.Add(emptyCluster);
+		}
+
+		protected abstract bool ClusterCondition(CPos p);
+
+		public bool IsConnected(CPos p1, CPos p2)
+		{
+			if (!clusterCellLayer.Contains(p1) || !clusterCellLayer.Contains(p2))
+				return false;
+
+			if (clusterCellLayer[p1] == clusterCellLayer[p2])
+				return true;
+
+			// Even though p1 and p2 are in different domains, it's possible
+			// that some dynamic terrain (i.e. bridges) may connect them.
+			return clusterCellLayer[p1].TransientConnections.Any(con => con.Active && clusterCellLayer[p2].TransientConnections.Contains(con));
+		}
+
+		public void UpdateCells(HashSet<CPos> dirtyCells)
+		{
+			var addedCells = dirtyCells.Where(cell => ClusterCondition(cell) && clusterCellLayer[cell].IsEmpty);
+			var removedCells = dirtyCells.Where(cell => !ClusterCondition(cell) && !clusterCellLayer[cell].IsEmpty);
+
+			foreach (var cell in addedCells)
+			{
+				var neighborClusters = new List<ClusterContents>();
+
+				// Select all neighbor clusters
+				var neighbors = CVec.Directions.Select(d => d + cell)
+					.Where(c => map.Contains(c) && !removedCells.Contains(c) && ClusterCondition(c));
+
+				ClusterContents foundCluster = null;
+				foreach (var n in neighbors)
+				{
+					if (foundCluster == null)
+					{
+						// Join the first viable neighbor we find.
+						foundCluster = clusterCellLayer[n];
+						UpdateCell(cell, clusterCellLayer[n]);
+					}
+					else
+					{
+						// Assimilate nearby neighbors
+						AssignCluster(clusterCellLayer[n], foundCluster);
+					}
+				}
+
+				if (foundCluster == null)
+				{
+					// If we can't join a neighbor, create a new cluster
+					UpdateCell(cell, new ClusterContents());
+					clusters.Add(clusterCellLayer[cell]);
+				}
+			}
+
+			foreach (var cell in removedCells)
+			{
+				// We don't need to test for splitting if it's the last cell in a domain
+				if (clusterCellLayer[cell].Cells.Count > 1)
+				{
+					// Select all neighbors inside the map boundaries
+					var neighbors = CVec.Directions.Select(d => d + cell)
+						.Where(c => clusterCellLayer[cell].Cells.Contains(c) && !removedCells.Contains(c)).ToList();
+
+					// We don't need to test for splitting if it's an edge of a domain
+					if (neighbors.Count >= 2)
+						RecalculateCluster(clusterCellLayer[cell], removedCells.ToList());
+				}
+
+				UpdateCell(cell, emptyCluster);
+			}
+		}
+
+		public void AddFixedConnection(IEnumerable<CPos> connectedCells)
+		{
+			var cells = connectedCells.ToList();
+			var connection = new ClusterTransientConnection(true, cells);
```

Comment:
TS bridge connections can be turned on and off. Temporarily they are all set to to true